kbs/src/api/src/token/coco.rs

```diff
@@ -55,6 +90,74 @@ impl AttestationTokenVerifier for CoCoAttestationTokenVerifier {
             }
         }
 
+        if let Some(trusted_store) = &self.trust_certs {
+            // Get certificate chain from 'x5c' or 'x5u' in JWK.
+            let cert_chain: Vec<X509> = match rsa_jwk.x5c {
+                Some(x5c) => {
+                    let mut cert_chain: Vec<X509> = vec![];
+                    for base64_der_cert in x5c {
+                        let der_cert = URL_SAFE_NO_PAD.decode(base64_der_cert)?;
+                        let cert = X509::from_der(&der_cert)?;
+                        cert_chain.push(cert)
+                    }
+                    cert_chain
+                }
+                None => {
+                    if let Some(x5u) = rsa_jwk.x5u {
+                        let res = reqwest::get(x5u).await?;
+                        match res.status() {
+                            reqwest::StatusCode::OK => {
+                                let pem_cert_chain = res.text().await?;
+                                let mut chain = Vec::new();
+
+                                for pem in pem_cert_chain.split("-----END CERTIFICATE-----") {
+                                    let trimmed =
+                                        format!("{}\n-----END CERTIFICATE-----", pem.trim());
+                                    if !trimmed.starts_with("-----BEGIN CERTIFICATE-----") {
+                                        continue;
+                                    }
+                                    let cert = X509::from_pem(trimmed.as_bytes())
+                                        .map_err(|_| anyhow!("Invalid PEM certificate chain"))?;
+                                    chain.push(cert);
+                                }
+                                chain
+                            }
+                            _ => {
+                                bail!(
+                                    "Request x5u in Attestation Token JWK Failed, Response: {:?}",
+                                    res.text().await?
+                                );
+                            }
+                        }
+                    } else {
+                        bail!("Missing certificate in Attestation Token JWK");
+                    }
+                }
+            };
+
+            // Check certificate is valid and trustworthy
+            let mut untrusted_stack = Stack::<X509>::new()?;
```

Comment:
Can't we push each cert onto this stack as part of the previous step? (rather than making a Vec<X509> and then iterating through it and cloning here)

Reply:
I have put the generation of `Vec<X509>` cert chain of `x5u` in a help function `download_cert_chain` now, so here iterating it and clone them (skip first cert) to Stack is OK for now.